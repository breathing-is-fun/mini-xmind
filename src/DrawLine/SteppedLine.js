import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import Line from './Line';

export default class SteppedLine extends PureComponent {
  static propTypes = {
    x0: PropTypes.number.isRequired,
    y0: PropTypes.number.isRequired,
    x1: PropTypes.number.isRequired,
    y1: PropTypes.number.isRequired,
    orientation: PropTypes.oneOf(['h', 'v']),
    borderColor: PropTypes.string,
    borderStyle: PropTypes.string,
    borderWidth: PropTypes.number,
    className: PropTypes.string,
    zIndex: PropTypes.number,
  };

  render() {
    if (this.props.orientation === 'h') {
      return this.renderHorizontal();
    }
    return this.renderVertical();
  }

  renderVertical() {
    const { x0, y0, x1, y1 } = this.props;

    const dx = x1 - x0;
    if (dx === 0) {
      return <Line {...this.props} />;
    }

    const borderWidth = this.props.borderWidth || defaultBorderWidth;
    const y2 = (y0 + y1) / 2;

    const xOffset = dx > 0 ? borderWidth : 0;
    const minX = Math.min(x0, x1) - xOffset;
    const maxX = Math.max(x0, x1);

    return (
      <div className="react-steppedlineto">
        <Line {...this.props} x0={x0} y0={y0} x1={x0} y1={y2} />
        <Line {...this.props} x0={x1} y0={y1} x1={x1} y1={y2} />
        <Line {...this.props} x0={minX} y0={y2} x1={maxX} y1={y2} />
      </div>
    );
  }

  renderHorizontal() {
    const { x0, y0, x1, y1 } = this.props;

    const dy = y1 - y0;
    if (dy === 0) {
      return <Line {...this.props} />;
    }

    const borderWidth = this.props.borderWidth || defaultBorderWidth;
    const x2 = (x0 + x1) / 2;

    const yOffset = dy < 0 ? borderWidth : 0;
    const minY = Math.min(y0, y1) - yOffset;
    const maxY = Math.max(y0, y1);

    return (
      <div className="react-steppedlineto">
        <Line {...this.props} x0={x0} y0={y0} x1={x2} y1={y0} />
        <Line {...this.props} x0={x1} y0={y1} x1={x2} y1={y1} />
        <Line {...this.props} x0={x2} y0={minY} x1={x2} y1={maxY} />
      </div>
    );
  }
}
