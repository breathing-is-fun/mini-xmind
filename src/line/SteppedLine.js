import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import Line from './core';
const defaultBorderWidth = 1;

export default class SteppedLine extends PureComponent {
  static propTypes = {
    orientation: PropTypes.oneOf(['h', 'v', 'horizonal', 'vertical']),
    className: PropTypes.string,
    style: PropTypes.object,
  };

  static defaultProps = {
    orientation: 'h',
    className: '',
    style: {},
  };

  renderVertical = () => {
    const { x0, y0, x1, y1 } = this.props;

    const dx = x1 - x0;
    if (dx === 0) {
      return <Line {...this.props} />;
    }

    const borderWidth = this.props.borderWidth || defaultBorderWidth;
    const y2 = (y0 + y1) / 2;

    const xOffset = dx > 0 ? borderWidth : 0;
    const minX = Math.min(x0, x1) - xOffset;
    const maxX = Math.max(x0, x1);

    return (
      <div className="stepped-lineto">
        <Line {...this.props} x0={x0} y0={y0} x1={x0} y1={y2} />
        <Line {...this.props} x0={x1} y0={y1} x1={x1} y1={y2} />
        <Line {...this.props} x0={minX} y0={y2} x1={maxX} y1={y2} />
      </div>
    );
  };

  renderHorizontal = () => {
    const { x0, y0, x1, y1 } = this.props;

    const dy = y1 - y0;
    if (dy === 0) {
      return <Line {...this.props} />;
    }

    const borderWidth = this.props.borderWidth || defaultBorderWidth;
    const x2 = (x0 + x1) / 2;

    const yOffset = dy < 0 ? borderWidth : 0;
    const minY = Math.min(y0, y1) - yOffset;
    const maxY = Math.max(y0, y1);

    return (
      <div className="stepped-line">
        <Line {...this.props} x0={x0} y0={y0} x1={x2} y1={y0} />
        <Line {...this.props} x0={x1} y0={y1} x1={x2} y1={y1} />
        <Line {...this.props} x0={x2} y0={minY} x1={x2} y1={maxY} />
      </div>
    );
  };

  render() {
    const { orientation } = this.props;
    let result;

    switch (orientation) {
      case 'horizonal':
      case 'h':
        result = this.renderHorizontal();
        break;

      case 'vertical':
      case 'v':
        result = this.renderVertical();
        break;
    }

    return result;
  }
}
