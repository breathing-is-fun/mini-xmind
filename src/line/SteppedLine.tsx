import React from 'react';
import Line from './Line';
import {
  getLineCoordinatesForHorizonal,
  getLineCoordinatesForVertical,
} from '../utils/LineUtil';

export type OrientationProps = 'horizonal' | 'vertical';
export interface SteppedLineProps {
  className?: string;
  from: any;
  to: any;
  offset?: { x: number; y: number };
  orientation?: OrientationProps;
}

const SteppedLine = ({
  from,
  to,
  offset = { x: 0, y: 0 },
  orientation,
}: SteppedLineProps) => {
  const renderHorizonal = () => {
    const {
      x0,
      x1,
      y0,
      y1,
      centerX,
      firstLineVisible,
      secondLineVisible,
      thirdLineVisible,
    } = getLineCoordinatesForHorizonal(from, to, offset);

    return (
      <div className="stepped-line">
        {firstLineVisible && <Line x0={x0} y0={y0} x1={centerX} y1={y0} />}
        {secondLineVisible && (
          <Line x0={centerX} y0={y0} x1={centerX} y1={y1} />
        )}
        {thirdLineVisible && <Line x0={centerX} y0={y1} x1={x1} y1={y1} />}
      </div>
    );
  };

  const renderVertical = () => {
    const {
      x0,
      x1,
      y0,
      y1,
      centerY,
      firstLineVisible,
      secondLineVisible,
      thirdLineVisible,
    } = getLineCoordinatesForVertical(from, to, offset);

    return (
      <div className="stepped-line">
        {firstLineVisible && <Line x0={x0} y0={y0} x1={x0} y1={centerY} />}
        {secondLineVisible && (
          <Line x0={x0} y0={centerY} x1={x1} y1={centerY} />
        )}
        {thirdLineVisible && <Line x0={x1} y0={centerY} x1={x1} y1={y1} />}
      </div>
    );
  };

  if (orientation === 'horizonal') {
    return renderHorizonal();
  }

  if (orientation === 'vertical') {
    return renderVertical();
  }

  return renderHorizonal();
};

export default SteppedLine;
