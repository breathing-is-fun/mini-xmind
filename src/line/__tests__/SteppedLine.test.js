import React from 'react';
import { mount } from 'enzyme';
import SteppedLine from '../SteppedLine';

describe('SteppedLine', () => {
  const getTargetIndexProps = (wrapper, name, index) =>
    wrapper
      .find(name)
      .at(index)
      .props();

  it('renderVertical should work well', () => {
    const wrapper = mount(
      <SteppedLine x0={1} y0={1} x1={0} y1={0} orientation="vertical" />,
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.line-core').length).toBe(3);
    expect(getTargetIndexProps(wrapper, 'Line', 0).orientation).toBe(
      'vertical',
    );
    expect(getTargetIndexProps(wrapper, '.line-core', 0).style.left).toBe(
      '1px',
    );
    expect(getTargetIndexProps(wrapper, '.line-core', 0).style.top).toBe('1px');
    expect(getTargetIndexProps(wrapper, '.line-core', 0).style.width).toBe(
      '0.5px',
    );

    expect(getTargetIndexProps(wrapper, 'Line', 1).orientation).toBe(
      'vertical',
    );
    expect(getTargetIndexProps(wrapper, '.line-core', 1).style.left).toBe(
      '0px',
    );
    expect(getTargetIndexProps(wrapper, '.line-core', 1).style.top).toBe('0px');
    expect(getTargetIndexProps(wrapper, '.line-core', 1).style.width).toBe(
      '0.5px',
    );

    expect(getTargetIndexProps(wrapper, 'Line', 2).orientation).toBe(
      'vertical',
    );
    expect(getTargetIndexProps(wrapper, '.line-core', 2).style.left).toBe(
      '0px',
    );
    expect(getTargetIndexProps(wrapper, '.line-core', 2).style.top).toBe(
      '0.5px',
    );
    expect(getTargetIndexProps(wrapper, '.line-core', 2).style.width).toBe(
      '1px',
    );
  });

  it('renderHorizontal should work well', () => {
    const wrapper = mount(
      <SteppedLine x0={1} y0={1} x1={0} y1={0} orientation="horizonal" />,
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.line-core').length).toBe(3);

    expect(getTargetIndexProps(wrapper, 'Line', 0).orientation).toBe(
      'horizonal',
    );
    expect(getTargetIndexProps(wrapper, '.line-core', 0).style.left).toBe(
      '1px',
    );
    expect(getTargetIndexProps(wrapper, '.line-core', 0).style.top).toBe('1px');
    expect(getTargetIndexProps(wrapper, '.line-core', 0).style.width).toBe(
      '0.5px',
    );

    expect(getTargetIndexProps(wrapper, 'Line', 1).orientation).toBe(
      'horizonal',
    );
    expect(getTargetIndexProps(wrapper, '.line-core', 1).style.left).toBe(
      '0px',
    );
    expect(getTargetIndexProps(wrapper, '.line-core', 1).style.top).toBe('0px');
    expect(getTargetIndexProps(wrapper, '.line-core', 1).style.width).toBe(
      '0.5px',
    );

    expect(getTargetIndexProps(wrapper, 'Line', 2).orientation).toBe(
      'horizonal',
    );
    expect(getTargetIndexProps(wrapper, '.line-core', 2).style.left).toBe(
      '0.5px',
    );
    expect(getTargetIndexProps(wrapper, '.line-core', 2).style.top).toBe(
      '-1px',
    );
    expect(getTargetIndexProps(wrapper, '.line-core', 2).style.width).toBe(
      '2px',
    );
  });

  it('orientation should work', () => {
    const wrapper = mount(
      <SteppedLine x0={1} y0={1} x1={0} y1={0} orientation="test" />,
    );
    expect(wrapper).toEqual({});
  });

  it('return one Line if x1 === x0', () => {
    const wrapper = mount(
      <SteppedLine x0={1} y0={1} x1={1} y1={0} orientation="v" />,
    );
    expect(wrapper.find('Line').length).toBe(1);
  });
});
