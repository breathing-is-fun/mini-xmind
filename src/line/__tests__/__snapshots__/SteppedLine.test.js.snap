// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SteppedLine render horizonal correctly 1`] = `
<SteppedLine
  from={
    Object {
      "bottom": 616,
      "height": 80,
      "left": 361,
      "right": 461,
      "top": 536,
      "width": 100,
      "x": 361,
      "y": 536,
    }
  }
  to={
    Object {
      "bottom": 428,
      "height": 80,
      "left": 65,
      "right": 165,
      "top": 348,
      "width": 100,
      "x": 65,
      "y": 348,
    }
  }
>
  <div
    className="stepped-line"
  >
    <Line
      x0={361}
      x1={263}
      y0={576}
      y1={576}
    >
      <div
        className="line-core"
        style={
          Object {
            "transform": "translate3d(361px, 576px, 0) rotate(180deg)",
            "width": "98px",
          }
        }
      />
    </Line>
    <Line
      x0={263}
      x1={263}
      y0={576}
      y1={388}
    >
      <div
        className="line-core"
        style={
          Object {
            "transform": "translate3d(263px, 576px, 0) rotate(-90deg)",
            "width": "188px",
          }
        }
      />
    </Line>
    <Line
      x0={263}
      x1={165}
      y0={388}
      y1={388}
    >
      <div
        className="line-core"
        style={
          Object {
            "transform": "translate3d(263px, 388px, 0) rotate(180deg)",
            "width": "98px",
          }
        }
      />
    </Line>
  </div>
</SteppedLine>
`;

exports[`SteppedLine render vertical correctly 1`] = `
<SteppedLine
  from={
    Object {
      "bottom": 616,
      "height": 80,
      "left": 361,
      "right": 461,
      "top": 536,
      "width": 100,
      "x": 361,
      "y": 536,
    }
  }
  orientation="vertical"
  to={
    Object {
      "bottom": 428,
      "height": 80,
      "left": 65,
      "right": 165,
      "top": 348,
      "width": 100,
      "x": 65,
      "y": 348,
    }
  }
>
  <div
    className="stepped-line"
  >
    <Line
      x0={411}
      x1={411}
      y0={536}
      y1={482}
    >
      <div
        className="line-core"
        style={
          Object {
            "transform": "translate3d(411px, 536px, 0) rotate(-90deg)",
            "width": "54px",
          }
        }
      />
    </Line>
    <Line
      x0={411}
      x1={115}
      y0={482}
      y1={482}
    >
      <div
        className="line-core"
        style={
          Object {
            "transform": "translate3d(411px, 482px, 0) rotate(180deg)",
            "width": "296px",
          }
        }
      />
    </Line>
    <Line
      x0={115}
      x1={115}
      y0={482}
      y1={428}
    >
      <div
        className="line-core"
        style={
          Object {
            "transform": "translate3d(115px, 482px, 0) rotate(-90deg)",
            "width": "54px",
          }
        }
      />
    </Line>
  </div>
</SteppedLine>
`;
